numcodecs/zfp.pyx

```diff
@@ -0,0 +1,422 @@
+ if not ret:
+    raise RuntimeError("zfp_read_header failed")
+
+ # determine destination data type 
+ if field._type == zfp_type_float:
+    the_type = np.float32 
+    type_size = 4 
+ elif field._type == zfp_type_double:
+    the_type = np.float64 
+    type_size = 8 
+ elif field._type == zfp_type_int32:
+    the_type = np.int32
+    type_size = 4 
+ else:
+    the_type = np.int64
+    type_size = 8 
+
+ # determine data dimension
+ nx,ny,nz,nw = field.nx,field.ny,field.nz,field.nw
+ datashape = (nx if nx != 0 else 1)* (ny if ny != 0 else 1) \
+           *(nz if nz != 0 else 1)* (nw if nw != 0 else 1)
+
+ # currently only serial exeution available
+ zfp_stream_set_execution(zfp,zfp_exec_serial)
+
+ # setup destination buffer
+ if dest is None:
+    dest = PyBytes_FromStringAndSize(NULL,datashape*type_size)
+    dest_ptr = PyBytes_AS_STRING(dest)
+ else:
+    arr = ensure_contiguous_ndarray(dest)
+    dest_buffer = Buffer(arr, PyBUF_ANY_CONTIGUOUS | PyBUF_WRITEABLE)
+    dest_ptr = dest_buffer.ptr
+ zfp_field_set_pointer(field,dest_ptr)
+
+ # perform decompression
+ ret = zfp_decompress(zfp,field)
+ if not ret: 
+    raise RuntimeError("decompress failed")
+
+ # release buffers
+ source_buffer.release()
+ if dest_buffer is not None:
+    dest_buffer.release()
+ zfp_field_free(field)
+ zfp_stream_close(zfp)
+ stream_close(stream)
+
+ # reshape the buffer
+ if not ny:
+   buf_shape=(nx)
+ elif ny and not nz:
+   buf_shape=(ny,nx)
+ elif ny and nz and not nw:
+   buf_shape=(nz,ny,nx)
+ else:
+   buf_shape=(nw,nz,ny,nx)
+ dest = ensure_ndarray(dest).view(the_type)
+ dest = dest.reshape(buf_shape)
+ return dest
+
+class CompressMethod:
+ '''
+ Create a CompressMethod object to store zfp mode and 
+ its parameters in Zfp class
+ 
+ '''
+
+ # Initilize with defaults
+ def __init__(self):
+    self.zfpmode='a'
+    self.tolerance=0.01
+
+ def set_mode(self,mode):
+     self.zfpmode=mode
+ def set_tolerance(self,tol):
+     self.tolerance=tol
+ def set_precision(self,prec):
+     self.precision=prec
+ def set_rate(self,rate):
+     self.rate=rate
+ def set_minbits(self,minbits):
+     self.minbits=minbits
+ def set_maxbits(self,maxbits):
+     self.maxbits=maxbits
+ def set_maxprec(self,maxprec):
+     self.maxprec=maxprec
+ def set_minexp(self,minexp):
+     self.minexp=minexp
+     
+class Zfp(Codec):
+ '''Codec providing compression using zfp via the Python standard library.
+ 
+ Parameters
+ ----------
+ mode : a
+ tolerance : float (absolue error tolerance), default 0.01
+ 
+ mode : p
+ precision : int (uncompressed bits per value), can be 0-64 for double precision
+
+ mode : r
+ rate : int (compressed bits per floating-point value), can be 0-64 for double precision
+
+ mode : c
+ minbits : int (min bits per 4^d values in d dimensions)
+ maxbits : int (max bits per 4^d values in d dimensions), 0 for unlimited
+ minexp : int (min bit plane : -1074 for all bit planes)
+ maxprec : int (max bits of precision per value), 0 for full
+ '''
```

Comment:
This docstring needs to follow [numpydoc conventions](https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes).